File: java/com/purescraps/purehtml/transformers/types/ResolveTransformer.java
package com.purescraps.purehtml.transformers.types;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

public class ResolveTransformer extends Transformer {
    @SuppressWarnings("unused")
    private List<String>  args;
    @SuppressWarnings("unused")
    public ResolveTransformer(List<String>  args) {
        this.args = args;
    }
    @SuppressWarnings("unused")
    public ResolveTransformer(){
        this.args = null;
    }

    public void setArgs(List<String> args){
        this.args = args;
    }
    // Static method to return the name of the transformer
    public String getName() {
        return "resolve";
    }

    // The transform method that resolves the URL
    public Object transform(TransformParams params) {
        String val = (String) params.getVal();
        String url = params.getUrl();  // Base URL from params

        if (val != null) {
            try {
                URI baseUri = new URI(url);
                URI resolvedUri = baseUri.resolve(val);  // Resolve the relative URL

                return resolvedUri.toString();  // Return the resolved URI as a string
            } catch (URISyntaxException e) {
                System.out.println(e.getMessage());
            }
        }
        System.out.println("ResolveTransformer.transform: invalid value type: " + (val == null ? "null" : val.getClass().getSimpleName()));
        System.exit(-1);
        return null;
    }
}

File: python/purehtml/transformers/types/resolve_transformer.py
from abc import ABC
from typing import List
from urllib.parse import urljoin

from purehtml.transformers.transformer import Transformer


class ResolveTransformer(Transformer, ABC):
    def __init__(self, args: List[str] = None):
        """
        Initialize the transformer with optional arguments.
        :param args: Arguments to configure the transformer (unused in this case).
        """
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        """
        Set the list of attributes to retrieve.
        :param args: A list of attribute names.
        """
        self.args = args

    def get_name(self) -> str:
        """
        Return the name of the transformer.
        :return: The name of the transformer.
        """
        return "resolve"

    def transform(self, params) -> str:
        """
        Resolves a relative URL based on a base URL.
        :param params: The parameters that contain the value (relative URL) and the base URL.
        :return: The resolved absolute URL.
        """
        val = params.get_val()  # Get the relative URL value from the params
        base_url = params.get_url()  # Get the base URL from the params

        if val is not None:
            try:
                # Resolve the relative URL to an absolute URL using urljoin
                resolved_url = urljoin(base_url, val)
                return resolved_url

            except Exception as e:
                print(f"Error resolving URL: {e}")
                raise

        raise ValueError(f"ResolveTransformer: Invalid value type: {type(val)}")
